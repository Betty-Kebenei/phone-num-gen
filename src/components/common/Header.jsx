import React, { Component } from 'react';

import binarySearch from './binarySearch';

class Header extends Component {
  state = {
    number: ''
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    const { number } = this.state;
    const allNumbers = localStorage.getItem('phonenumbers');
    if(allNumbers !== null){
      const allNumbersArr = JSON.parse(allNumbers)
      const numbers = allNumbersArr.sort();
      const response = await binarySearch(numbers, number);
      if(response) {
        // eslint-disable-next-line no-alert
        alert('Cheers!!! That number exists!')
      } else {
        // eslint-disable-next-line no-alert
        alert('We are sorry!!! That number DOES NOT exists!')
      }
    } else {
      // eslint-disable-next-line no-alert
      alert('We are sorry!!! You have no numbers to search!')
    }
    this.setState({number: ''})
  }

  render(){
    const { number } = this.state;
    const currentRoute = window.location.pathname;
    return(
      <div className="navbar navbar-expand-lg navbar-dark bg-dark">
        <button
          className="navbar-toggler"
          type="button"
          data-toggle="collapse"
          data-target="#collapsed-navbar"
          aria-controls="collapsed-navbar"
          aria-expanded="false"
          aria-label="Toggle navigation">
          <span className="navbar-toggler-icon" />
        </button>
        <a className="navbar-brand" href="/">Phone Number Generator</a>
        <div className="collapse navbar-collapse" id="collapsed-navbar">
          <ul className="navbar-nav mr-auto">
            <li className={currentRoute === '/' ? "nav-item active" : "nav-item"}>
              <a className="nav-link" href="/">Generate Number</a>
            </li>
            <li className={currentRoute === '/numbers' ? "nav-item active" : "nav-item"}>
              <a className="nav-link" href="/numbers">Phone Numbers</a>
            </li>
          </ul>
          <form className="form-inline my-2 my-lg-0" onSubmit={this.handleSubmit}>
            <input
              className="form-control mr-sm-2"
              type="search"
              name="search"
              aria-label="Search"
              placeholder="search number"
              value={number}
              onChange={event => this.setState({number: event.target.value})}
              />
            <button className="btn btn-outline-success my-2 my-sm-0" type="submit">Search</button>
          </form>
        </div>
      </div>
    );
  }
}
export default Header;
