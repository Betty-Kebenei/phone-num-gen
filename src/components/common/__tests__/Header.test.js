import React from 'react';
import renderer from 'react-test-renderer';
import {configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon'

import Header from '../Header';

configure({ adapter: new Adapter() })
window.alert = jest.fn();

describe('Header', () => {
  test('header component renders correctly', () => {
    const component = renderer.create(<Header />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('handleSubmit: search when localstorage has content', () => {
    const arr = [
      "0520824292",
      "0324483933",
      "0219081561",
      "0305923108",
      "0240259478",
      "0421493267",
      "0607632178",
      "0882948791",
      "0010663129",
      "0246821435"]

    test('existing number found if searched', async() => {
      localStorage.setItem('phonenumbers', JSON.stringify(arr));
      const wrapper = mount(<Header />);
      wrapper.setState({number: '0246821435'});
      const renderedComponent = wrapper.instance();
      sinon.spy(renderedComponent, 'handleSubmit');
      await renderedComponent.handleSubmit({preventDefault: () => {}});
      expect(window.alert).toHaveBeenCalledWith('Cheers!!! That number exists!')
      expect(wrapper.state().number).toEqual('');
    })

    test('non-existing number not found if searched', async() => {
      localStorage.setItem('phonenumbers', JSON.stringify(arr));
      const wrapper = mount(<Header />);
      wrapper.setState({number: '0987654320'});
      const renderedComponent = wrapper.instance();
      sinon.spy(renderedComponent, 'handleSubmit');
      await renderedComponent.handleSubmit({preventDefault: () => {}});
      expect(window.alert).toHaveBeenCalledWith('We are sorry!!! That number DOES NOT exists!')
      expect(wrapper.state().number).toEqual('');
    })
  })

  describe('handleSubmit: search when localstorage has no content', () => {
    test('nothing is searched', () => {
      localStorage.clear();
      const wrapper = mount(<Header />);
      wrapper.setState({number: ''});
      wrapper.find('[name="search"]').simulate('change', {target: {name: 'search', value: '0987654320'}});
      const renderedComponent = wrapper.instance();
      sinon.spy(renderedComponent, 'handleSubmit');
      renderedComponent.handleSubmit({preventDefault: () => {}});
      expect(window.alert).toHaveBeenCalledWith('We are sorry!!! You have no numbers to search!')
      expect(wrapper.state().number).toEqual('');
    })
  })
})
