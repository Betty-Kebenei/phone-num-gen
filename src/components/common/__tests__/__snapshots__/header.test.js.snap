// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header header component renders correctly 1`] = `
<div
  className="navbar navbar-expand-lg navbar-dark bg-dark"
>
  <button
    aria-controls="collapsed-navbar"
    aria-expanded="false"
    aria-label="Toggle navigation"
    className="navbar-toggler"
    data-target="#collapsed-navbar"
    data-toggle="collapse"
    type="button"
  >
    <span
      className="navbar-toggler-icon"
    />
  </button>
  <a
    className="navbar-brand"
    href="/"
  >
    Phone Number Generator
  </a>
  <div
    className="collapse navbar-collapse"
    id="collapsed-navbar"
  >
    <ul
      className="navbar-nav mr-auto"
    >
      <li
        className="nav-item active"
      >
        <a
          className="nav-link"
          href="/"
        >
          Generate Number
        </a>
      </li>
      <li
        className="nav-item"
      >
        <a
          className="nav-link"
          href="/numbers"
        >
          Phone Numbers
        </a>
      </li>
    </ul>
    <form
      className="form-inline my-2 my-lg-0"
      onSubmit={[Function]}
    >
      <input
        aria-label="Search"
        className="form-control mr-sm-2"
        name="search"
        onChange={[Function]}
        placeholder="search number"
        type="search"
        value=""
      />
      <button
        className="btn btn-outline-success my-2 my-sm-0"
        type="submit"
      >
        Search
      </button>
    </form>
  </div>
</div>
`;
